### PR TITLE
Improve Errors

### LINKED ISSUE #100: Give taffy::Error a more useful name
Monolithic error types like this are rarely useful, and it make reading the code hard.

### PR DESCRIPTION
**Pull request by TheDestroyer19**
# Objective

Replaces `taffy::Error` with `taffy::NodeNotFoundError` to make code much more legible. Also adds `taffy::ChildOperationError` so that `remove_child_at_index`, `replace_child_at_index`, and `child_at_index` can return an error instead of panicing.

Fixes #100 #104

## Feedback wanted

I'm not happy with the name of `ChildOperationError`. It feels a little vague to me. Is it fine, or is there possibly a better name for it?

Also, I'm not sure I should have flattened NodeNotFoundError into a tuple struct.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -29,27 +29,62 @@ pub use crate::node::Taffy;
...
+    ChildIndexOutOfBounds {
+        /// The parent node whose child was being looked up
+        parent: node::Node,
+        /// The index that was looked up
+        child_index: usize,
+        /// The total number of children the parent has
+        child_count: usize,
+    },
+    ///The [`Node`](node::Node) was not found in the [`Taffy`] instance
+    NodeNotFoundError(NodeNotFoundError),
```
Comment (alice-i-cecile) [inline src/lib.rs:59]: IMO we can just keep this as an enum variant; no need for a seperate type.
```suggestion
    NodeNotFound(node::Node),
```

File: src/lib.rs [hunk 2]
```diff
@@ -29,27 +29,62 @@ pub use crate::node::Taffy;
...
+        write!(f, "Node {:?} is not in the Taffy instance", self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for NodeNotFoundError {}
+
+/// An error that occurs while trying to access or modify a [`Node`](node::Node)'s children by index.
+#[derive(Debug)]
+pub enum ChildOperationError {
```
Comment (alice-i-cecile) [inline src/lib.rs:48]:
```suggestion
pub enum NodeError {
```

I think that a single unified "errors that occur when working with Nodes" is a solid straightforward design here.

File: src/lib.rs [hunk 3]
```diff
@@ -29,27 +29,62 @@ pub use crate::node::Taffy;
...
+impl Display for NodeNotFoundError {
+    fn fmt(&self, f: &mut Formatter) -> Result {
+        write!(f, "Node {:?} is not in the Taffy instance", self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for NodeNotFoundError {}
+
+/// An error that occurs while trying to access or modify a [`Node`](node::Node)'s children by index.
```
Comment (alice-i-cecile) [inline src/lib.rs:46]: Update these docs to match the new design :)

File: src/lib.rs [hunk 4]
```diff
@@ -29,27 +29,62 @@ pub use crate::node::Taffy;
...
-/// An error that can occur when performing layout
+/// The [`Node`](node::Node) was not found in the [`Taffy`] instance
 #[derive(Debug)]
-pub enum Error {
-    /// The [`Node`](node::Node) was invalid
-    InvalidNode(node::Node),
+pub struct NodeNotFoundError(pub node::Node);
+
+#[cfg(feature = "std")]
+impl Display for NodeNotFoundError {
```
Comment (alice-i-cecile) [inline src/lib.rs:37]: I think we should just pull in `thiserror` and feature-gate it behind `std`. It's a much nicer way to define these error strings.

File: src/lib.rs [hunk 5]
```diff
@@ -29,27 +29,62 @@ pub use crate::node::Taffy;
 #[cfg(feature = "std")]
 use core::fmt::{Display, Formatter, Result};
 
-/// An error that can occur when performing layout
+/// The [`Node`](node::Node) was not found in the [`Taffy`] instance
 #[derive(Debug)]
-pub enum Error {
-    /// The [`Node`](node::Node) was invalid
-    InvalidNode(node::Node),
+pub struct NodeNotFoundError(pub node::Node);
```
Comment (alice-i-cecile) [inline src/lib.rs:34]: I like this as a tuple struct; there's only one possibility here.

File: src/lib.rs (hunk 2 above)
Comment (alice-i-cecile) [inline src/lib.rs:48]: Hmm. I don't love this name either. Maybe just `ChildError`?

File: src/lib.rs (hunk 1 above)
Comment (alice-i-cecile) [inline src/lib.rs:59]: I don't love nesting these errors like this, but until we have "enum variants are types" this is a better design to keep the function signatures of most of the operations cleaner.

File: src/lib.rs (hunk 2 above)
Comment (alice-i-cecile) [reply src/lib.rs:48]: Actually: what if we went with `InvalidNode` and `InvalidChild`? And then stuck these in an `error` module, so then users can call them with the path as `error::InvalidNode` if they want?

This gets my vote :)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Thanks! First round of feedback is up :) Some suggestions on naming mostly.

I'd like feedback on whether or not we should use `thiserror`.

The API is much nicer, but it [pulls in](https://crates.io/crates/thiserror-impl/1.0.24/dependencies) some relatively heavy proc-macro crates which might hurt compile times.

File: src/lib.rs (hunk 2 above)
Comment (TheDestroyer19) [reply src/lib.rs:48]: The `error` module idea sounds good. Thank you for the names.

File: RELEASES.md [hunk 6]
```diff
@@ -4,14 +4,16 @@
 
 ### 0.2.0 Added
 
-- Nothing yet
+- added `taffy::ChildOperationError` enum
```
Comment (alice-i-cecile) [inline RELEASES.md:7]: This is outdated.

File: RELEASES.md [hunk 7]
```diff
@@ -4,14 +4,16 @@
...
 
 ### 0.2.0 Changed
 
 - `Sprawl::remove` now returns a `Result<usize, Error>`, to indicate if the operation was sucessful, and if it was, which ID was invalidated.
 - renamed `taffy::forest::Forest.new-node(..)` `taffy::forest::Forest.new_with_children(..)`
 - renamed `taffy::node::Taffy.new-node(..)` -> `taffy::node::Taffy.new_with_children(..)`
 - renamed `taffy::style::Style` -> `taffy::style::FlexboxLayout` to more precicely indicate its purpose
+- renamed `taffy::Error` -> `taffy::NodeNotFoundError` and converted it into a tuple struct to precisely show what causes it
```
Comment (alice-i-cecile) [inline RELEASES.md:15]: Also outdated :)

File: src/error.rs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    /// The parent [`Node`](node::Node) does not have a child at `child_index`. It only has `child_count` children
+    ChildIndexOutOfBounds {
+        /// The parent node whose child was being looked up
+        parent: Node,
+        /// The index that was looked up
+        child_index: usize,
+        /// The total number of children the parent has
+        child_count: usize,
+    },
+    /// The [`Node`](node::Node) was not found in the [`Taffy`] instance.
```
Comment (alice-i-cecile) [inline src/error.rs:33]:
```suggestion
    /// The parent [`Node`](node::Node) was not found in the [`Taffy`] instance.
```

Comment (alice-i-cecile) [reply src/error.rs:33]: I believe this is correct?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Almost ready now.

File: src/error.rs (hunk 8 above)
Comment (TheDestroyer19) [reply src/error.rs:33]: yes we can - `Taffy::replace_child_at_index()` if `new_child` isn't valid.

**Comment by alice-i-cecile** [comment]
Excellent work; thanks for being so responsive to reviews.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by TheDestroyer19** [comment]
Thank you for being quick to provide feedback. I appreciate it

**Pull request merged**